src/client/executor.rs

```diff
@@ -309,167 +285,152 @@ impl Client {
         &self,
         mut op: T,
         mut session: Option<&mut ClientSession>,
-    ) -> Result<ExecutionOutput<T>> {
-        // If the current transaction has been committed/aborted and it is not being
-        // re-committed/re-aborted, reset the transaction's state to TransactionState::None.
-        if let Some(ref mut session) = session {
-            if matches!(
-                session.transaction.state,
-                TransactionState::Committed { .. }
-            ) && op.name() != CommitTransaction::NAME
-                || session.transaction.state == TransactionState::Aborted
-                    && op.name() != AbortTransaction::NAME
-            {
-                session.transaction.reset();
+    ) -> Result<ExecutionDetails<T>> {
+        let mut retry: Option<ExecutionRetry> = None;
+        let mut implicit_session: Option<ClientSession> = None;
+        loop {
+            if retry.is_some() {
+                op.update_for_retry();
             }
-        }
-
-        let selection_criteria = session
-            .as_ref()
-            .and_then(|s| s.transaction.pinned_mongos())
-            .or_else(|| op.selection_criteria());
 
-        let server = match self.select_server(selection_criteria).await {
-            Ok(server) => server,
-            Err(mut err) => {
-                err.add_labels_and_update_pin(None, &mut session, None)?;
-                return Err(err);
+            // If the current transaction has been committed/aborted and it is not being
+            // re-committed/re-aborted, reset the transaction's state to TransactionState::None.
+            if let Some(ref mut session) = session {
+                if matches!(
+                    session.transaction.state,
+                    TransactionState::Committed { .. }
+                ) && op.name() != CommitTransaction::NAME
+                    || session.transaction.state == TransactionState::Aborted
+                        && op.name() != AbortTransaction::NAME
+                {
+                    session.transaction.reset();
+                }
             }
-        };
 
-        let mut conn = match get_connection(&session, &op, &server.pool).await {
-            Ok(c) => c,
-            Err(mut err) => {
-                err.add_labels_and_update_pin(None, &mut session, None)?;
-                if err.is_read_retryable() && self.inner.options.retry_writes != Some(false) {
-                    err.add_label(RETRYABLE_WRITE_ERROR);
-                }
+            let selection_criteria = session
+                .as_ref()
+                .and_then(|s| s.transaction.pinned_mongos())
+                .or_else(|| op.selection_criteria());
```

Comment:
ditto above, does this need to be done in the loop?

Reply:
Unfortunately not - this (potentially) borrows from `op`, so moving it outside the loop would block the call to `op.update_for_retry()` (which requires a &mut).